agent/hcp/manager.go

```diff
@@ -71,15 +77,80 @@ func NewManager(cfg ManagerConfig) *Manager {
 	}
 }
 
+// runReporter initializes, and if successful, then runs the metrics reporter.
+func (m *Manager) runReporter(ctx context.Context) error {
+	// Step 1: Obtain CCM telemetry configuration
+	// Only enable HCP metrics reporting if server is registered with management plane.
+	telemetryCfg, err := m.cfg.Client.FetchTelemetryConfig(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to obtain CCM telemetry config: %v", err)
+	}
+
+	if telemetryCfg == nil {
+		// Not an error, as the server is not registered with CCM, but early return.
+		return nil
+	}
+
+	// Step 2: Init telemetry.MetricsExporter which sends metrics to HCP Metrics Gateway in OTLP format.
+	// It uses a an OTLP exporter client wrapped by the HCP client.
+	// This enables us to perform HCP auth and easily mock the client interface for tests.
+	// It must first be initialized within the HCP Client with the configured endpoint.
+	if err := m.cfg.Client.InitMetricsClient(ctx, telemetryCfg.Endpoint); err != nil {
+		return fmt.Errorf("failed to init metrics HCP client: %v", err)
+	}
+
+	expCfg := &telemetry.MetricsExporterConfig{
+		Labels: map[string]string{
+			"service.name":        "consul-server",
+			"service.version":     version.GetHumanVersion(),
+			"service.instance.id": m.cfg.NodeID,
+		},
+		Logger:  m.logger.Named("metrics_exporter"),
+		Filters: telemetryCfg.Filters,
+		// Inject client in metrics exporter.
+		Client: m.cfg.Client,
+	}
+
+	exp, err := telemetry.NewMetricsExporter(expCfg)
+	if err != nil {
+		return fmt.Errorf("failed to create exporter: %v", err)
+	}
+
+	// Step 3: Init telemetry.Reporter, which gathers consul server go metrics over a configurable time interval (Report Interval).
+	// It flushes them to the exporter to be sent to HCP at a configurable time interval (Batch Interval).
+	cfg := telemetry.DefaultConfig()
+	cfg.Logger = m.logger.Named("telemetry_reporter")
+	cfg.Gatherer = m.cfg.MetricsBackend
+	cfg.Exporter = exp
+
+	m.reporter, err = telemetry.NewReporter(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create exporter: %v", err)
+	}
+
+	// If setup is successful, run the reporter, which is a blocking operation.
+	m.reporter.Run(ctx)
+
+	return nil
```

Comment:
Could also make the `m.reporter.Run(ctx)` return an error to make this `return m.reporter.Run(ctx)`, but the reporter can be run outside of this package, and so it felt forced to do that.